examples/bigquery_user_info_updater_tool/bigquery_user_info_updater/updater_tools/query_creator.py

```diff
@@ -0,0 +1,336 @@
+                    field,
+                    record_name,
+                    record_alias
+                )
+                field_str = inner_string.format(
+                    name,
+                    alias
+                )
+                inner_string_list.append(field_str)
+
+        def _process_select_string_field(
+                field,
+                record_alias=None
+        ):
+            """Creates a list of fields to be selected as part of the
+                gather_updates query by recursively processing fields and nested
+                fields.
+
+            Args:
+                field(dict): Represents a field in the user schema.
+                record_alias(str): The alias of the record that the field
+                    belongs to if the field variable is a nested field. None
+                    if the field is not a nested field.
+                        """
+            if field['type'] == 'RECORD':
+                struct_str = 'STRUCT({0:s}) as {1:s}'
+                sub_field_list = []
+                for sub_field in field['fields']:
+                    alias_string = '{0:s}_{1:s}'
+                    if record_alias:
+                         alias = alias_string.format(
+                             record_alias,
+                             sub_field['name']
+                         )
+                    else:
+                        alias = alias_string.format(
+                            field['name'],
+                            sub_field['name']
+                        )
+                    sub_field_list.append(_process_select_string_field(
+                        sub_field,
+                        alias
+                    ))
+                sub_field_str = ', '.join(sub_field_list)
+
+                return struct_str.format(
+                    sub_field_str,
+                    field['name']
+                )
+            else:
+                if record_alias:
+                    return '{0:s} as {1:s}'.format(
+                        record_alias,
+                        field['name']
+                    )
+                else:
+                    return field['name']
+
+        # Get the templates to build the queries
+        abs_path = os.path.abspath(os.path.dirname(__file__))
+        outer_string_path = os.path.join(
+            abs_path,
+            '../query_templates/gather_updates_outer_string.txt')
+        with open(outer_string_path, 'r') as input_file:
+            outer_string = input_file.read()
+
+        inner_string_list = []
+        field_list = []
+        inner_string_path = os.path.join(
+            abs_path,
+            '../query_templates/gather_updates_inner_string.txt')
+        with open(inner_string_path, 'r') as input_file:
+            inner_string = input_file.read()
+
+        # Transform the user schema into a dict for processing
+        with open(self.schema_path, 'r') as f:
```

Comment:
jaketf on Aug 27, 2019 Member
use with context to close file when done with it.